File: solutions/r/hamming/hamming.R
# tokenize returns a vector with all of the characters from a string.
tokenize <- function(string) {
    strsplit(string, "")[[1]]
}

# This function takes two strings and calculates the hamming distance.
hamming <- function(strand1, strand2) {
    if (nchar(strand1) != nchar(strand2)) {
        stop("different lengths")
    }
    sum(tokenize(strand1) != tokenize(strand2))
}

File: solutions/ruby/hamming/hamming.rb
module Hamming
  def self.compute(strand1, strand2)
    raise ArgumentError unless strand1.length == strand2.length
    strand1.chars.each_with_index.reduce(0) { |distance, (c, i)|
      distance + (c != strand2[i] ? 1 : 0)
    }
  end
end
